File: 3297/main.go
package main

import (
	"fmt"
)

func main() {
	fmt.Println(validSubstringCount("bcca", "abc"))
}

func validSubstringCount(word1 string, word2 string) int64 {
	count := make([]int, 26)
	for _, c := range word2 {
		count[c-'a']++
	}
	n := len(word1)
	preCount := make([][]int, n+1)
	for i := range preCount {
		preCount[i] = make([]int, 26)
	}
	for i := 1; i <= n; i++ {
		copy(preCount[i], preCount[i-1])
		preCount[i][word1[i-1]-'a']++
	}
	var res int64
	for l := 1; l <= n; l++ {
		r := get(l, n+1, preCount, count)
		res += int64(n-r+1)
	}
	return res
}

func get(l, r int, preCount [][]int, count []int) int {
	border := l
	for l < r {
		m := (l+r) >> 1
		f := true
		for i := 0; i < 26; i++ {
			if preCount[m][i]-preCount[border-1][i] < count[i] {
				f = false
				break
			}
		}
		if f {
			r = m
		} else {
			l = m + 1
		}
	}
	return l
}

File: 3297/main.rs
struct Solution {}


impl Solution {
    pub fn valid_substring_count(word1: String, word2: String) -> i64 {
        let mut diff = vec![0; 26];
        for c in word2.chars() {
            diff[(c as u8 - b'a') as usize] -= 1;
        }
        let mut res = 0;
        let mut cnt = diff.iter().filter(|&&c| c < 0).count();
        let mut update = |c: usize, add: i32, cnt: &mut usize | {
            diff[c] += add;
            if add == 1 && diff[c] == 0 {
                *cnt -= 1;
            } else if add == -1 && diff[c] == -1 {
                *cnt += 1;
            }
        };
        let (mut l, mut r) = (0, 0);
        let n = word1.len();
        let bytes = word1.as_bytes();
        while l < n {
            while r < n && cnt > 0 {
                update((bytes[r] - b'a') as usize, 1, &mut cnt);
                r += 1;
            }
            if cnt == 0 {
                res += (n - r) as i64 + 1;
            }
            update((bytes[l] - b'a') as usize, -1, &mut cnt);
            l += 1;
        }
        res
    }
}

fn main() {
    println!("{:#?}", Solution::valid_substring_count("bcca".to_string(), "abc".to_string()));
}